File: src/main/upgates.ts
import { getSetting, setSetting } from './db';
import { encrypt, decrypt } from './secure';
import { UpgatesOrder, UpgatesConfig, OrderLive } from '../shared/types';

/**
 * Upgates API (https://docs.upgates.com/api/intro)
 * - Basic auth (login API uživatele + klíč), URL tvaru https://ESHOP.admin.SERVER.upgates.com
 * - používáme jen čtení objednávek dle e-mailu zákazníka
 */

export function getUpgatesConfig(): UpgatesConfig {
  return {
    url: getSetting('upgatesUrl', '')!,
    login: getSetting('upgatesLogin', '')!,
    hasKey: !!getSetting('upgatesKey')
  };
}

export function saveUpgatesConfig(cfg: { url?: string; login?: string; apiKey?: string }): UpgatesConfig {
  if (cfg.url !== undefined) setSetting('upgatesUrl', cfg.url.trim().replace(/\/+$/, ''));
  if (cfg.login !== undefined) setSetting('upgatesLogin', cfg.login.trim());
  if (cfg.apiKey !== undefined) setSetting('upgatesKey', cfg.apiKey ? encrypt(cfg.apiKey) : '');
  ordersCache.clear();
  return getUpgatesConfig();
}

export function upgatesConfigured(): boolean {
  const c = getUpgatesConfig();
  return !!(c.url && c.login && c.hasKey);
}

async function upgatesFetch(pathAndQuery: string): Promise<any> {
  const cfg = getUpgatesConfig();
  const key = getSetting('upgatesKey');
  if (!cfg.url || !cfg.login || !key) throw new Error('Upgates API není nastaveno (Nastavení → AI → Upgates)');
  const auth = Buffer.from(`${cfg.login}:${decrypt(key)}`).toString('base64');
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 15_000);
  try {
    const res = await fetch(`${cfg.url}${pathAndQuery}`, {
      headers: { Authorization: `Basic ${auth}`, 'Content-Type': 'application/json' },
      redirect: 'follow',
      signal: controller.signal
    });
    if (res.status === 401 || res.status === 403) throw new Error('Upgates API odmítlo přihlášení — zkontroluj login a klíč');
    if (!res.ok) throw new Error(`Upgates API: HTTP ${res.status}`);
    return await res.json();
  } finally {
    clearTimeout(timer);
  }
}

export async function testUpgates(): Promise<string> {
  const data = await upgatesFetch('/api/v2/status');
  const allowed = (data?.services ?? []).filter((s: any) => s.privilege !== 'deny').length;
  return `Připojení funguje (${allowed} povolených endpointů).`;
}

function mapOrder(o: any): UpgatesOrder {
  return {
    orderNumber: o.order_number ?? '',
    status: o.status ?? '',
    creationTime: o.creation_time ?? '',
    paidDate: o.paid_date ?? null,
    deliveredDate: o.delivered_date ?? null,
    trackingCode: o.tracking_code ?? null,
    trackingUrl: o.tracking_url ?? null,
    total: o.order_total ?? 0,
    currency: o.currency_id ?? '',
    shipmentName: o.shipment?.name ?? '',
    paymentName: o.payment?.name ?? '',
    products: (o.products ?? [])
      .filter((p: any) => p.type !== 'shipment' && p.type !== 'payment')
      .map((p: any) => `${p.quantity}× ${p.title}`)
      .slice(0, 8),
    adminUrl: o.admin_url ?? null
  };
}

// Cache 10 minut, ať šetříme rate limit
const ordersCache = new Map<string, { at: number; orders: UpgatesOrder[] }>();

export async function ordersByEmail(emailRaw: string): Promise<UpgatesOrder[]> {
  const email = emailRaw.trim().toLowerCase();
  if (!email.includes('@')) return [];
  const hit = ordersCache.get(email);
  if (hit && Date.now() - hit.at < 10 * 60_000) return hit.orders;
  const data = await upgatesFetch(
    `/api/v2/orders?email=${encodeURIComponent(email)}&order_by=creation_time&order_dir=desc`
  );
  const orders = (data?.orders ?? []).map(mapOrder).slice(0, 10);
  ordersCache.set(email, { at: Date.now(), orders });
  return orders;
}

/**
 * Živý stav jedné objednávky dle čísla. Zkusí přímý dotaz na číslo objednávky,
 * a když ho API nepodporuje, dohledá objednávku mezi objednávkami zákazníka.
 */
export async function orderLive(orderNumber: string, email: string | null): Promise<OrderLive | null> {
  const wanted = orderNumber.replace(/^0+/, '');
  const same = (n: string) => n.replace(/^0+/, '') === wanted;

  let order: UpgatesOrder | null = null;
  try {
    const data = await upgatesFetch(`/api/v2/orders?order_number=${encodeURIComponent(orderNumber)}`);
    const hit = (data?.orders ?? []).find((o: any) => same(String(o.order_number ?? '')));
    if (hit) order = mapOrder(hit);
  } catch { /* endpoint filtr nemusí být povolený — jdeme přes e-mail */ }

  if (!order && email) {
    const list = await ordersByEmail(email);
    order = list.find(o => same(o.orderNumber)) ?? null;
  }
  if (!order) return null;

  return {
    status: order.status || null,
    paid: !!order.paidDate,
    paidDate: order.paidDate,
    deliveredDate: order.deliveredDate,
    trackingCode: order.trackingCode,
    trackingUrl: order.trackingUrl,
    adminUrl: order.adminUrl
  };
}

/** Textový blok pro AI — poslední objednávky zákazníka jako zdroj faktů. */
export async function ordersContextForAi(email: string): Promise<string> {
  if (!upgatesConfigured()) return '';
  try {
    const orders = await ordersByEmail(email);
    if (orders.length === 0) return '';
    return orders.slice(0, 3).map(o => {
      const bits = [
        `Objednávka č. ${o.orderNumber} (${new Date(o.creationTime).toLocaleDateString('cs-CZ')})`,
        `stav: ${o.status || 'neuveden'}`,
        o.paidDate ? `zaplaceno ${o.paidDate}` : 'nezaplaceno',
        o.deliveredDate ? `doručeno ${o.deliveredDate}` : null,
        `doprava: ${o.shipmentName || '—'}`,
        o.trackingCode ? `tracking: ${o.trackingCode}${o.trackingUrl ? ` (${o.trackingUrl})` : ''}` : 'tracking zatím není',
        `celkem ${o.total} ${o.currency}`,
        `položky: ${o.products.join(', ') || '—'}`
      ].filter(Boolean);
      return `- ${bits.join(' | ')}`;
    }).join('\n');
  } catch {
    return '';
  }
}

File: ios/QuentinoApp/Services/Upgates.swift
import Foundation

/**
 Upgates API — živá data e-shopu.

 Používá se jen čtení objednávek podle e-mailu zákazníka. Přihlašuje se
 uživatelem API a klíčem (basic auth); klíč leží v klíčence, ne v databázi.
 Odpovědi se drží deset minut v paměti, aby se nešahalo na limit dotazů
 pokaždé, když se otevře stejná zpráva.
 */
enum Upgates {
    private static var cache: [String: (at: Date, orders: [[String: Any]])] = [:]
    private static let cacheQueue = DispatchQueue(label: "cz.quentino.upgates")

    static func config() -> [String: Any] {
        [
            "url": Store.setting("upgatesUrl", "") ?? "",
            "login": Store.setting("upgatesLogin", "") ?? "",
            "hasKey": Secrets.has("upgatesKey")
        ]
    }

    static func saveConfig(_ patch: [String: Any]) -> [String: Any] {
        if let url = patch["url"] as? String {
            Store.setSetting("upgatesUrl", url.trimmedSlash)
        }
        if let login = patch["login"] as? String {
            Store.setSetting("upgatesLogin", login.trimmingCharacters(in: .whitespaces))
        }
        if let key = patch["apiKey"] as? String, !key.isEmpty { Secrets.set("upgatesKey", key) }
        cacheQueue.sync { cache.removeAll() }
        Store.touchState()
        return config()
    }

    static var isReady: Bool {
        let current = config()
        return !(current["url"] as? String ?? "").isEmpty
            && !(current["login"] as? String ?? "").isEmpty
            && (current["hasKey"] as? Bool ?? false)
    }

    private static func call(_ pathAndQuery: String) async throws -> [String: Any] {
        let current = config()
        let url = current["url"] as? String ?? ""
        let login = current["login"] as? String ?? ""
        let key = Secrets.get("upgatesKey") ?? ""
        guard !url.isEmpty, !login.isEmpty, !key.isEmpty else {
            throw BridgeError.message("Upgates API není nastaveno (Nastavení → AI → Upgates).")
        }
        let auth = Data("\(login):\(key)".utf8).base64EncodedString()
        do {
            return try await Http.dictionary(
                "\(url)\(pathAndQuery)",
                headers: ["Authorization": "Basic \(auth)", "Content-Type": "application/json"],
                timeout: 20
            )
        } catch let failure as Http.Failure {
            if failure.status == 401 || failure.status == 403 {
                throw BridgeError.message("Upgates API odmítlo přihlášení — zkontroluj login a klíč.")
            }
            throw BridgeError.message("Upgates API: HTTP \(failure.status)")
        }
    }

    static func test() async throws -> String {
        let data = try await call("/api/v2/status")
        let allowed = (data["services"] as? [[String: Any]] ?? [])
            .filter { ($0["privilege"] as? String) != "deny" }.count
        return "Připojení funguje (\(allowed) povolených endpointů)."
    }

    private static func order(_ raw: [String: Any]) -> [String: Any] {
        let products = (raw["products"] as? [[String: Any]] ?? [])
            .filter { ($0["type"] as? String) != "shipment" && ($0["type"] as? String) != "payment" }
            .prefix(8)
            .map { "\($0["quantity"] ?? 1)× \($0["title"] as? String ?? "")" }
        return [
            "orderNumber": raw["order_number"] as? String ?? "",
            "status": raw["status"] as? String ?? "",
            "creationTime": raw["creation_time"] as? String ?? "",
            "paidDate": raw["paid_date"] ?? NSNull(),
            "deliveredDate": raw["delivered_date"] ?? NSNull(),
            "trackingCode": raw["tracking_code"] ?? NSNull(),
            "trackingUrl": raw["tracking_url"] ?? NSNull(),
            "total": raw["order_total"] ?? 0,
            "currency": raw["currency_id"] as? String ?? "",
            "shipmentName": (raw["shipment"] as? [String: Any])?["name"] as? String ?? "",
            "paymentName": (raw["payment"] as? [String: Any])?["name"] as? String ?? "",
            "products": Array(products),
            "adminUrl": raw["admin_url"] ?? NSNull()
        ]
    }

    /**
     Objednávky zákazníka tak, jak je vrátí API.

     Karta objednávky z nich potřebuje víc než zjednodušený přehled (položky
     s variantami, adresy, ceny dopravy), proto se drží obojí a mapuje se až
     na výstupu.
     */
    static func rawOrders(email rawEmail: String) async throws -> [[String: Any]] {
        let email = rawEmail.trimmingCharacters(in: .whitespaces).lowercased()
        guard email.contains("@") else { return [] }
        if let hit = cacheQueue.sync(execute: { cache[email] }), Date().timeIntervalSince(hit.at) < 600 {
            return hit.orders
        }
        let data = try await call(
            "/api/v2/orders?email=\(Http.escaped(email))&order_by=creation_time&order_dir=desc"
        )
        let out = Array((data["orders"] as? [[String: Any]] ?? []).prefix(10))
        cacheQueue.sync { cache[email] = (Date(), out) }
        return out
    }

    static func orders(email rawEmail: String) async throws -> [[String: Any]] {
        try await rawOrders(email: rawEmail).map(order)
    }

    /// Textový blok pro AI — poslední objednávky zákazníka jako zdroj faktů.
    static func contextForAi(email: String) async -> String {
        guard isReady, let list = try? await orders(email: email), !list.isEmpty else { return "" }
        return list.prefix(3).map { order -> String in
            var bits = ["Objednávka č. \(order["orderNumber"] as? String ?? "")"]
            bits.append("stav: \((order["status"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? "neuveden")")
            bits.append((order["paidDate"] as? String).map { "zaplaceno \($0)" } ?? "nezaplaceno")
            if let delivered = order["deliveredDate"] as? String { bits.append("doručeno \(delivered)") }
            bits.append("doprava: \((order["shipmentName"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? "—")")
            if let tracking = order["trackingCode"] as? String {
                bits.append("tracking: \(tracking)")
            } else {
                bits.append("tracking zatím není")
            }
            bits.append("celkem \(order["total"] ?? 0) \(order["currency"] as? String ?? "")")
            let products = (order["products"] as? [String] ?? []).joined(separator: ", ")
            bits.append("položky: \(products.isEmpty ? "—" : products)")
            return "- \(bits.joined(separator: " | "))"
        }.joined(separator: "\n")
    }
}
